File: ios/Beacons/Models/FirmwareManifest.swift
import Foundation
import Combine

/// The firmware manifest published at soyboi.tech. It tells the app the latest version
/// per board (keyed by the exact "fw" label a board reports), whether that board can take
/// an in-app OTA update, and where to download the verified image (or which browser flasher
/// to send the user to when OTA isn't offered).
///
/// Shape (schema 1):
/// {
///   "schema": 1,
///   "updated": "YYYY-MM-DD",
///   "builds": {
///     "beacon board": {
///       "version": "2.0.0",
///       "ota": true,
///       "app": { "url": "https://...", "sha256": "<hex64 or empty>", "size": <int> },
///       "flasher": "https://...",
///       "notes": ""
///     }, ...
///   }
/// }
///
/// The manifest is advisory: a bad/missing/malformed fetch never blocks the UI or crashes.
/// We fall back to the last-good cache, then to a baked-in default that mirrors
/// DeviceStatus.latestVersion, so the "update available" nudge always has an answer.
struct FirmwareManifest: Codable, Equatable {
    var schema: Int
    var updated: String?
    var builds: [String: Build]

    struct Build: Codable, Equatable {
        var version: String
        var ota: Bool
        var app: AppImage
        var flasher: String?
        var notes: String?
        /// Absent on every non-beacon board: only the dual-radio build ships a co-processor.
        var nrf: NrfImage?
    }

    /// The co-processor half of a build entry. The nRF52840 runs the Adafruit/Seeed bootloader,
    /// which speaks LEGACY Nordic DFU (service 0x1530), so what ships is an adafruit-nrfutil DFU
    /// .zip, not a raw image. `version` is a small monotonic Int (NRF_APP_VERSION in the nRF
    /// firmware), NOT a semver string: the S3 reports the running value as "nrfv" and an update
    /// is offered when the manifest's is higher.
    struct NrfImage: Codable, Equatable {
        var version: Int
        var ota: Bool
        var url: String
        var sha256: String
        var size: Int
        /// Same rule as AppImage.sig: empty/absent means unsigned, so the update must be refused.
        var sig: String?
    }

    struct AppImage: Codable, Equatable {
        var url: String
        var sha256: String
        var size: Int
        // Detached ECDSA P-256/SHA-256 signature over the whole image, lowercase hex DER.
        // Empty/absent = unsigned = OTA must be refused (the board gate rejects it too).
        var sig: String?
    }
}

extension FirmwareManifest.Build {
    /// True when this entry carries a complete, verifiable image: an integrity hash, a real
    /// size, a URL, and a non-empty image signature. Part of the OTA gate; without all four
    /// we can't verify a download, and an unsigned image would be rejected by the board anyway.
    var hasVerifiableImage: Bool {
        !app.sha256.isEmpty && app.size > 0 && !app.url.isEmpty && !(app.sig ?? "").isEmpty
    }

    /// Same bar for the co-processor package. nil `nrf` (every non-beacon board) is false, so a
    /// caller can ask without checking for the sub-entry first.
    var hasVerifiableNrfImage: Bool {
        guard let n = nrf else { return false }
        return !n.sha256.isEmpty && n.size > 0 && !n.url.isEmpty && !(n.sig ?? "").isEmpty
    }
}

extension FirmwareManifest {
    /// The build entry for a board's fw label ("beacon board", "ACAB-ouispy",
    /// "mesh-detect-ACAB"), or nil if the manifest doesn't list it.
    func build(forFwLabel label: String) -> Build? { builds[label] }

    /// Baked-in fallback used when we've never fetched a manifest and have nothing cached.
    /// Mirrors the app's per-board ship targets (DeviceStatus.latestVersion for the beacon board,
    /// colonelLatestVersion for the single-board builds) so the offline nudge matches.
    /// OTA is off here on purpose: with no verified image details we only ever point the
    /// user at the browser flasher until a real manifest arrives.
    static var fallback: FirmwareManifest {
        let img = AppImage(url: "", sha256: "", size: 0, sig: nil)
        // Per-board flasher pages (matches the Android fallback): the beacon board flashes
        // from soyboi.tech, the Colonel Panic boards from the ACAB repo's flasher.
        let acabFlasher = "https://soyboi1312.github.io/all-cameras-are-beacons/"
        func build(_ version: String, _ flasher: String) -> Build {
            Build(version: version, ota: false, app: img,
                  flasher: flasher, notes: nil)
        }
        // The beacon board has moved ahead of the Colonel Panic single-board builds, so each
        // label carries its own offline baseline (matches the Android per-board fallback).
        return FirmwareManifest(schema: 1, updated: nil, builds: [
            "beacon board": build(DeviceStatus.latestVersion, "https://soyboi.tech/flash.html"),
            // rev-B rides the beacon version line but flashes from its own page: its image must
            // never land on rev-A hardware, so the fallback must not point it at flash.html.
            "beacon board rev-B": build(DeviceStatus.latestVersion, "https://soyboi.tech/flash-revb.html"),
            "ACAB-ouispy": build(DeviceStatus.colonelLatestVersion, acabFlasher),
            "mesh-detect-ACAB": build(DeviceStatus.colonelLatestVersion, acabFlasher),
            "mesh-detect-ACAB-ch1": build(DeviceStatus.colonelLatestVersion, acabFlasher),
        ])
    }
}

/// Fetches, caches, and serves the firmware manifest. Injected as an environment object so
/// the Device screen can ask "what's the latest for this board, and can it self-update?".
///
/// Fetch policy: non-blocking background refresh on app start and on board connect, with a
/// ~6 h TTL. The last-good JSON is persisted to UserDefaults with its fetch timestamp and
/// used verbatim when offline. Nothing here ever blocks the UI or throws into it.
@MainActor
final class FirmwareManifestStore: ObservableObject {
    static let shared = FirmwareManifestStore()

    /// The manifest the UI should read. Starts from cache (or the baked-in fallback) and is
    /// replaced in place by a successful refresh.
    @Published private(set) var manifest: FirmwareManifest

    private let manifestURL = URL(string: "https://soyboi.tech/firmware/firmware-latest.json")!
    private let cacheKey = "acab.firmwareManifest"       // last-good JSON
    private let cacheTimeKey = "acab.firmwareManifestAt" // when we fetched it (epoch seconds)
    private let ttl: TimeInterval = 6 * 60 * 60          // refresh at most once per ~6 h
    private nonisolated static let maxManifestBytes = 256 * 1024
    private var inFlight = false

    private init() {
        // Seed from the last-good cache, else the baked-in fallback. Either way the UI has a
        // usable answer before any network call finishes.
        if let cached = Self.readCache(key: cacheKey) {
            manifest = cached
        } else {
            manifest = .fallback
        }
    }

    // MARK: Accessors the UI consumes

    /// Latest known version for a board's fw label. Falls back to DeviceStatus.latestVersion
    /// when the manifest doesn't list the board.
    func latestVersion(forFwLabel label: String) -> String {
        manifest.build(forFwLabel: label)?.version ?? DeviceStatus.latestVersion
    }

    /// The full manifest entry for a board's fw label, or nil when unlisted.
    func entry(forFwLabel label: String) -> FirmwareManifest.Build? {
        manifest.build(forFwLabel: label)
    }

    // MARK: Refresh

    /// Kick a background refresh if the cache is stale (or forced). Safe to call often: it
    /// no-ops when a fetch is in flight or the cache is still fresh, and it never blocks or
    /// surfaces an error to the caller.
    func refreshIfNeeded(force: Bool = false) {
        guard !inFlight else { return }
        if !force, let at = UserDefaults.standard.object(forKey: cacheTimeKey) as? Double,
           Date().timeIntervalSince1970 - at < ttl {
            return   // cache still within TTL
        }
        inFlight = true
        Task { await self.fetch() }
    }

    /// Force an immediate refresh and await it, bypassing the TTL. Backs the manual
    /// "check for updates" control so the UI can show a spinner until it resolves and then
    /// re-evaluate update availability off the freshly-published manifest. No-ops (and
    /// returns right away) when a background refresh is already in flight.
    func refreshNow() async {
        guard !inFlight else { return }
        inFlight = true
        await fetch()
    }

    private func fetch() async {
        defer { inFlight = false }
        var req = URLRequest(url: manifestURL)
        req.cachePolicy = .reloadIgnoringLocalCacheData   // we do our own TTL/caching
        req.timeoutInterval = 15
        guard let payload = try? await Self.boundedManifestData(for: req),
              (200..<300).contains(payload.response.statusCode),
              let fetched = try? JSONDecoder().decode(FirmwareManifest.self, from: payload.data),
              fetched.schema == 1, !fetched.builds.isEmpty else {
            // Bad network, non-2xx, or malformed/unsupported JSON: keep serving whatever we
            // already have. Never clobber a good cache with garbage.
            return
        }
        let data = payload.data
        UserDefaults.standard.set(data, forKey: cacheKey)
        UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: cacheTimeKey)
        manifest = fetched
    }

    private nonisolated static func boundedManifestData(
        for request: URLRequest
    ) async throws -> (data: Data, response: HTTPURLResponse) {
        guard let url = request.url, url.scheme?.lowercased() == "https",
              url.host?.lowercased() == "soyboi.tech", url.user == nil,
              url.password == nil, url.port == nil || url.port == 443 else {
            throw FirmwareManifestFetchError.invalidResponse
        }
        let configuration = URLSessionConfiguration.ephemeral
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        let session = URLSession(
            configuration: configuration,
            delegate: FirmwareManifestRejectRedirectsDelegate(),
            delegateQueue: nil
        )
        defer { session.finishTasksAndInvalidate() }
        let (bytes, response) = try await session.bytes(for: request)
        guard let http = response as? HTTPURLResponse,
              http.url?.scheme?.lowercased() == "https",
              http.url?.host?.lowercased() == "soyboi.tech",
              http.expectedContentLength <= Int64(maxManifestBytes) else {
            throw FirmwareManifestFetchError.invalidResponse
        }
        var data = Data()
        if http.expectedContentLength > 0 {
            data.reserveCapacity(min(maxManifestBytes, Int(http.expectedContentLength)))
        }
        for try await byte in bytes {
            guard data.count < maxManifestBytes else {
                throw FirmwareManifestFetchError.tooLarge
            }
            data.append(byte)
        }
        return (data, http)
    }

    private static func readCache(key: String) -> FirmwareManifest? {
        guard let data = UserDefaults.standard.data(forKey: key),
              let m = try? JSONDecoder().decode(FirmwareManifest.self, from: data) else { return nil }
        return m
    }
}

private enum FirmwareManifestFetchError: Error {
    case invalidResponse
    case tooLarge
}

private final class FirmwareManifestRejectRedirectsDelegate: NSObject, URLSessionTaskDelegate,
    @unchecked Sendable {
    nonisolated func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

File: android/app/src/main/java/tech/acab/app/net/FirmwareManifest.kt
package tech.acab.app.net

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.atomic.AtomicBoolean

/** Read an HTTP body with both declared-length and streaming ceilings. Null means refusal. */
internal fun readBoundedManifestBody(
    input: InputStream,
    declaredLength: Long,
    maxBytes: Int,
): ByteArray? {
    if (declaredLength > maxBytes || maxBytes <= 0) return null
    val out = ByteArrayOutputStream(
        if (declaredLength in 1..maxBytes.toLong()) declaredLength.toInt() else minOf(maxBytes, 16 * 1024),
    )
    val buf = ByteArray(8 * 1024)
    var total = 0
    while (true) {
        val read = input.read(buf)
        if (read < 0) break
        total += read
        if (total > maxBytes) return null
        out.write(buf, 0, read)
    }
    return out.toByteArray()
}

/** Firmware artifacts are intentionally confined to the product's exact HTTPS origin. */
internal fun trustedFirmwareArtifactUrl(raw: String): URL? = runCatching {
    URL(raw).takeIf { url ->
        url.protocol.equals("https", ignoreCase = true) &&
            url.host.equals("soyboi.tech", ignoreCase = true) &&
            url.userInfo == null && url.ref == null && url.port in setOf(-1, 443)
    }
}.getOrNull()

/** Redirects are refused even when their destination would otherwise be trusted. */
internal fun firmwareArtifactResponseAllowed(initial: URL, final: URL, statusCode: Int): Boolean =
    statusCode == HttpURLConnection.HTTP_OK &&
        trustedFirmwareArtifactUrl(initial.toExternalForm()) != null &&
        trustedFirmwareArtifactUrl(final.toExternalForm()) != null &&
        initial.toExternalForm() == final.toExternalForm()

/**
 * The co-processor half of a build entry, from builds[<label>].nrf. The nRF52840 runs the
 * Adafruit/Seeed bootloader, which speaks LEGACY Nordic DFU (service 0x1530), so what the
 * manifest publishes is an adafruit-nrfutil DFU .zip rather than a raw image. [version] is a
 * small monotonic int (NRF_APP_VERSION in the nRF firmware), NOT a semver string: the S3
 * reports the running value as "nrfv" and an update is offered when the manifest's is higher.
 */
data class NrfBuild(
    val version: Int,
    val ota: Boolean,
    val zipUrl: String,
    val sha256: String,
    val size: Long,
    val sig: String,
) {
    /** Same publication bar as the S3 image: a download URL, a published hash, a bounded size,
     *  and a signature. The app verifies the detached signature because the stock nRF bootloader
     *  cannot, so an unsigned package must never reach the hardware. */
    val hasVerifiableImage: Boolean get() =
        zipUrl.isNotEmpty() && sha256.isNotEmpty() && size in 1L..MAX_PACKAGE_BYTES && sig.isNotEmpty()

    companion object {
        const val MAX_PACKAGE_BYTES = 4L * 1024 * 1024
    }
}

/**
 * One firmware build's entry from the manifest. Keys mirror the JSON at
 * https://soyboi.tech/firmware/firmware-latest.json under builds[<fw label>].
 */
data class FirmwareBuild(
    val version: String,
    val ota: Boolean,
    val appUrl: String,
    val sha256: String,
    val size: Long,
    val sig: String,
    val flasher: String,
    val notes: String,
    /** null when the entry publishes no co-processor package (every non-beacon board). */
    val nrf: NrfBuild? = null,
    /** Exact builds-map key that selected this image. Empty only for the non-OTA fallback. */
    val manifestLabel: String = "",
) {
    /** True when the entry carries a downloadable, verifiable image: a non-empty download
     *  URL, a non-empty hash, a real size, and a non-empty ECDSA signature (hex DER over the
     *  image's SHA-256). A blank URL or hash, zero size, or a missing signature means "no
     *  verifiable image published yet", so in-app OTA is never offered off this entry even
     *  when ota == true. An unsigned image would be refused by the board's signature gate,
     *  so we never start such an update. */
    val hasVerifiableImage: Boolean get() =
        appUrl.isNotEmpty() && sha256.isNotEmpty() && size > 0 && sig.isNotEmpty()
}

/** The whole parsed manifest: a schema version, a fetched-at stamp, and one entry per fw label. */
data class FirmwareManifestData(
    val schema: Int,
    val updated: String,
    val builds: Map<String, FirmwareBuild>,
) {
    fun build(forFwLabel: String?): FirmwareBuild? =
        forFwLabel?.let { builds[it] }
}

/**
 * Fetches and caches the firmware manifest. The manifest is the source of truth for
 * "which version is current" and "can this board update in-app". Fetching is non-blocking
 * and failure-tolerant: a bad network or malformed JSON leaves the last-good cache (or the
 * baked-in fallback) in place, and never throws to the caller.
 *
 * Process-wide singleton so one fetch backs every screen. Collect [manifest] from the UI.
 */
class FirmwareManifest private constructor(context: Context) {

    private val prefs = context.applicationContext
        .getSharedPreferences("acab.firmware", Context.MODE_PRIVATE)

    private val _manifest = MutableStateFlow(loadCachedOrFallback())
    val manifest: StateFlow<FirmwareManifestData> = _manifest.asStateFlow()

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val fetching = AtomicBoolean(false)

    /** The current entry for a board's fw label (null when the manifest has none). */
    fun build(forFwLabel: String?): FirmwareBuild? = _manifest.value.build(forFwLabel)

    /** The current published version for a board's fw label, or null when unknown. Callers
     *  compare this to the installed version with their own numeric comparator. */
    fun latestVersion(forFwLabel: String?): String? =
        _manifest.value.build(forFwLabel)?.version

    /**
     * Refresh in the background if the cache is stale (older than [ttlMs]) or forced. Returns
     * immediately; the [manifest] flow updates if a good fetch lands. Safe to call on app
     * start and on every board connect; a fetch already in flight is not duplicated.
     */
    fun refresh(force: Boolean = false) {
        val age = System.currentTimeMillis() - prefs.getLong(KEY_FETCHED_AT, 0L)
        if (!force && age < TTL_MS) return
        if (!fetching.compareAndSet(false, true)) return
        scope.launch {
            try {
                val body = withContext(Dispatchers.IO) { httpGet(MANIFEST_URL) }
                val parsed = body?.let { runCatching { parse(it) }.getOrNull() }
                if (parsed != null) {
                    // Persist the raw body (last-good) and stamp it, then publish.
                    prefs.edit()
                        .putString(KEY_JSON, body)
                        .putLong(KEY_FETCHED_AT, System.currentTimeMillis())
                        .apply()
                    _manifest.value = parsed
                }
                // On any failure we simply keep the existing value: cache or fallback.
            } finally {
                fetching.set(false)
            }
        }
    }

    /**
     * Force a fetch past the TTL and suspend until it resolves, for the manual "check for
     * updates" control. Failure-tolerant like [refresh]: a bad network or malformed JSON
     * leaves the last-good cache (or fallback) in place and never throws. The [manifest] flow
     * updates in place if a good fetch lands, so callers can just re-read it afterward.
     */
    suspend fun refreshNow() {
        // Skip if a background fetch is already in flight; its result lands on the same flow.
        if (!fetching.compareAndSet(false, true)) return
        try {
            val body = withContext(Dispatchers.IO) { httpGet(MANIFEST_URL) }
            val parsed = body?.let { runCatching { parse(it) }.getOrNull() }
            if (parsed != null) {
                prefs.edit()
                    .putString(KEY_JSON, body)
                    .putLong(KEY_FETCHED_AT, System.currentTimeMillis())
                    .apply()
                _manifest.value = parsed
            }
        } finally {
            fetching.set(false)
        }
    }

    private fun loadCachedOrFallback(): FirmwareManifestData {
        val raw = prefs.getString(KEY_JSON, null)
        if (raw != null) runCatching { return parse(raw) }
        return FALLBACK
    }

    companion object {
        private const val MANIFEST_URL = "https://soyboi.tech/firmware/firmware-latest.json"
        private const val KEY_JSON = "manifestJson"
        private const val KEY_FETCHED_AT = "manifestFetchedAt"
        private const val TTL_MS = 6L * 60 * 60 * 1000   // ~6 hours
        private const val MAX_MANIFEST_BYTES = 256 * 1024

        @Volatile private var INSTANCE: FirmwareManifest? = null

        fun getInstance(context: Context): FirmwareManifest =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: FirmwareManifest(context.applicationContext).also { INSTANCE = it }
            }

        /**
         * Baked-in fallback, used only when there is no cache and no network yet. Mirrors the
         * hardcoded LATEST the app shipped with, so the "update available" nudge still works
         * offline. Deliberately marks nothing OTA-capable and carries no image, so a cold,
         * offline start can only ever point at the browser flasher, never at an unverified
         * in-app download.
         */
        private val FALLBACK = FirmwareManifestData(
            schema = 1,
            updated = "",
            builds = mapOf(
                "beacon board" to FirmwareBuild(
                    version = "2.0.5", ota = false, appUrl = "", sha256 = "", size = 0L, sig = "",
                    flasher = "https://soyboi.tech/flash.html", notes = "",
                ),
                // rev-B rides the beacon version line but flashes from its own page: its image
                // must never land on rev-A hardware, so the fallback must not point it at
                // flash.html. Matches the iOS fallback.
                "beacon board rev-B" to FirmwareBuild(
                    version = "2.0.5", ota = false, appUrl = "", sha256 = "", size = 0L, sig = "",
                    flasher = "https://soyboi.tech/flash-revb.html", notes = "",
                ),
                "ACAB-ouispy" to FirmwareBuild(
                    version = "2.0.5", ota = false, appUrl = "", sha256 = "", size = 0L, sig = "",
                    flasher = "https://soyboi1312.github.io/all-cameras-are-beacons/", notes = "",
                ),
                "mesh-detect-ACAB" to FirmwareBuild(
                    version = "2.0.5", ota = false, appUrl = "", sha256 = "", size = 0L, sig = "",
                    flasher = "https://soyboi1312.github.io/all-cameras-are-beacons/", notes = "",
                ),
                "mesh-detect-ACAB-ch1" to FirmwareBuild(
                    version = "2.0.5", ota = false, appUrl = "", sha256 = "", size = 0L, sig = "",
                    flasher = "https://soyboi1312.github.io/all-cameras-are-beacons/", notes = "",
                ),
            ),
        )

        /** Parse the manifest JSON. Throws on structurally-bad input (caught by the caller),
         *  including a schema we don't speak or an empty builds map - mirroring the iOS
         *  adoption gate, so a schema-2 or truncated publish never displaces the last-good
         *  cache here either. */
        private fun parse(raw: String): FirmwareManifestData {
            val root = JSONObject(raw)
            require(root.optInt("schema", 0) == 1) { "unsupported manifest schema" }
            val buildsObj = root.getJSONObject("builds")
            val builds = HashMap<String, FirmwareBuild>()
            val keys = buildsObj.keys()
            while (keys.hasNext()) {
                val label = keys.next()
                val b = buildsObj.getJSONObject(label)
                val app = b.optJSONObject("app")
                builds[label] = FirmwareBuild(
                    version = b.optString("version", ""),
                    ota = b.optBoolean("ota", false),
                    appUrl = app?.optString("url", "") ?: "",
                    sha256 = app?.optString("sha256", "")?.lowercase() ?: "",
                    size = app?.optLong("size", 0L) ?: 0L,
                    sig = app?.optString("sig", "")?.lowercase() ?: "",
                    flasher = b.optString("flasher", ""),
                    notes = b.optString("notes", ""),
                    nrf = b.optJSONObject("nrf")?.let { n ->
                        NrfBuild(
                            version = n.optInt("version", 0),
                            ota     = n.optBoolean("ota", false),
                            zipUrl  = n.optString("url", ""),
                            sha256  = n.optString("sha256", "").lowercase(),
                            size    = n.optLong("size", 0L),
                            sig     = n.optString("sig", "").lowercase(),
                        )
                    },
                    manifestLabel = label,
                )
            }
            require(builds.isNotEmpty()) { "manifest carries no builds" }
            return FirmwareManifestData(
                schema = 1,
                updated = root.optString("updated", ""),
                builds = builds,
            )
        }

        /** Plain HTTPS GET on the caller's thread. Returns the body, or null on any error. */
        private fun httpGet(url: String): String? {
            var conn: HttpURLConnection? = null
            return try {
                conn = (URL(url).openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    connectTimeout = 8_000
                    readTimeout = 8_000
                    setRequestProperty("Accept", "application/json")
                }
                if (conn.responseCode != HttpURLConnection.HTTP_OK) return null
                val bytes = conn.inputStream.use { input ->
                    readBoundedManifestBody(input, conn.contentLengthLong, MAX_MANIFEST_BYTES)
                } ?: return null
                bytes.toString(Charsets.UTF_8)
            } catch (_: Exception) {
                null
            } finally {
                conn?.disconnect()
            }
        }
    }
}
